service/shardmanager/service.go

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) 2019 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package shardmanager
+
+import (
+	"sync/atomic"
+
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/dynamicconfig"
+	"github.com/uber/cadence/common/resource"
+	"github.com/uber/cadence/common/service"
+	"github.com/uber/cadence/service/shardmanager/config"
+)
+
+// Service represents the shard manager service
+type Service struct {
+	resource.Resource
+
+	status int32
+	//  handler handler.Handler
+	stopC  chan struct{}
+	config *config.Config
+}
+
+// NewService builds a new task manager service
+func NewService(
+	params *resource.Params,
+	factory resource.ResourceFactory,
+) (resource.Resource, error) {
+
+	serviceConfig := config.NewConfig(
+		dynamicconfig.NewCollection(
+			params.DynamicConfig,
+			params.Logger,
+			dynamicconfig.ClusterNameFilter(params.ClusterMetadata.GetCurrentClusterName()),
+		),
+		params.HostName,
+	)
+
+	serviceResource, err := factory.NewResource(
+		params,
+		service.ShardManager,
+		&service.Config{
+			PersistenceMaxQPS:        serviceConfig.PersistenceMaxQPS,
+			PersistenceGlobalMaxQPS:  serviceConfig.PersistenceGlobalMaxQPS,
+			ThrottledLoggerMaxRPS:    serviceConfig.ThrottledLogRPS,
+			IsErrorRetryableFunction: common.IsServiceTransientError,
+			// shard manager doesn't need visibility config as it never read or write visibility
+		},
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Service{
+		Resource: serviceResource,
+		status:   common.DaemonStatusInitialized,
+		config:   serviceConfig,
+		stopC:    make(chan struct{}),
+	}, nil
+}
+
+// Start starts the service
+func (s *Service) Start() {
+	if !atomic.CompareAndSwapInt32(&s.status, common.DaemonStatusInitialized, common.DaemonStatusStarted) {
+		return
+	}
+
+	logger := s.GetLogger()
+	logger.Info("shard manager starting")
+
+	// setup the handler
+
+	s.Resource.Start()
+
+	<-s.stopC
```

Comment:
What is this line for?
Awe we blocking on `Start()` call?

Reply:
Good question, I think it's a pretty questionable design, it's the same for all the services, e.g. matching (but also all the others):  
https://github.com/uber/cadence/blob/6108a443a82b7998582fc8870e012a21ae0a353c/service/matching/service.go#L85

We cannot just remove it, this code depends on it blocking: https://github.com/uber/cadence/blob/fb1324877829f126af9d7e9760636240a1f92bcc/cmd/server/cadence/server.go#L340

I don't know why we track the life cycle in two different places. I think the right way to fix is migrate to fx.